### PR TITLE
Change webpack to rollup and convert the code to es

### PR DESCRIPTION
**Pull request by emilos**
This happened to be a little bit more changes than I expected but here's an overview of changes:

1. eslint config was updated to handle `import export`
2. the `dist` was added to `.gitignore` and it was replaced with `prepublish: "npm run build"` in `package.json` instead so that the files are there after `npm publish` without being checked in
3. webpack + plugins were replaced with rollup + plugins
4. karma config was updated to use rollup
5. package.json got updated so that `nyc + karma` work with `import` syntax
6. `should` got replaced with `assert` because it was hard to set it up with rollup (I hope that's not a blocker)
7. `main` field in package json was replaced to `dist/...umd.js`
8. `module` field was added for `es` version (to be more friendly with bundlers)

uff that's it :)

All of the tests are passing and the output is much cleaner than it was before. It also works in older browsers now.

The reason for all of this was that the `main` field in package.json was pointing at the unprocessed version, which is totally fine for node, but might break in browsers. I wasn't aware of that and I used it in the browser which happened to break IE etc.

I hope that the above makes sense, thanks again :)

### CODE REVIEW

**Comment by tandrewnichols** [comment]
Hmm, this is a lot more than I expected. Which isn't all bad, but as a primarily-node developer, I cringe at the thought of requiring transpiled code. I would personally like to keep the node version requirable as is, which means import/export won't work. On the other hand, I understand why you've made the change to make main point to dist. It's pretty typical to exclude node modules in babel config when importing stuff, and that allows you to still `require('indefinite')` or `import indefinite from 'indefinite'`, as opposed to needing to remember to `import indefinite from 'indefinite/dist/indefinite'`. Is it a common isomorphic pattern to make the main file a transpiled one? The only "happy medium" solution I can think of is to make main point to some other file that does environment checking and returns the right file _for_ you. Seems kinda dirty.

I'm also wondering what the blocker was with should. I like should significantly more than assert. Like I said, I've never used rollup, so I don't know immediately what the problem was, but it also seems like it _should_ work fine. Did `require('should')` or `import 'should'` not work?

File: karma.conf.js [hunk 1]
```diff
@@ -9,33 +13,40 @@ module.exports = function(config) {
...
     browsers: ['Chrome', 'PhantomJS'],
     preprocessors: {
-      'test/**/*.js': ['webpack']
+      'test/**/*.js': ['rollup']
     },
 
     // list of files / patterns to load in the browser
     files: [
-      'test/**/*.js'
+      { pattern: 'test/**/*.js', watched: false }
```
Comment (tandrewnichols) [inline karma.conf.js]: What's the reasoning for `watched: false`? `gulp ci` watches, so won't this make it not restart when the tests change?

File: rollup.build.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import babel from 'rollup-plugin-babel'
+
+export default {
+  input: 'lib/indefinite.js',
+  output: [
```
Comment (tandrewnichols) [inline rollup.build.js]: Nice, thanks for taking this little extra bit on (I mean, it looks like it was pretty straight forward, but . . . .still not nothing).

File: test/indefinite.js [hunk 3]
```diff
@@ -1,119 +1,119 @@
...
+import assert from 'assert';
+import indefinite from '../dist/indefinite.cjs';
 
 describe('indefinite', () => {
-  const indefinite = require('../lib/indefinite');
 
   context('a word that starts with a vowel', () => {
     it('should be prefixed with an', () => {
-      indefinite('apple').should.equal('an apple');
+      assert(indefinite('apple') === 'an apple');
```
Comment (tandrewnichols) [inline test/indefinite.js]: I could live with this if I had too, but I'd like to try to see if we can get should to work.

File: rollup.build.js (hunk 2 above)
Comment (tandrewnichols) [reply rollup.build.js]: I don't know how rollup works . . . will it automatically create minified versions of each of these?

Comment (tandrewnichols) [reply rollup.build.js]: Also, we'll need to update the README to at least mention support for various formats.

File: package.json [hunk 4]
```diff
@@ -2,11 +2,15 @@
   "name": "indefinite",
   "description": "Prefix a noun with an indefinite article - a or an - based on whether it begins with a vowel",
   "version": "2.0.3",
-  "main": "lib/indefinite.js",
+  "main": "dist/indefinite.umd.js",
+  "module": "dist/indefinite.es.js",
```
Comment (tandrewnichols) [inline package.json]: I'm not familiar with this field. What is it for?

**Comment by tandrewnichols** [comment]
You were not wrong about should. What a pain. If I'm being honest, it doesn't give me a lot of faith in rollup. Requiring a module seems like it should be more straight forward than that, especially considering should even includes the same kind of structure (generated by rollup) that you're proposing, with a `cjs` directory and an `es6` directory.

That being said, this works (with only one warning about not exporting a default):

```js
import should from '../node_modules/should/should.js';
```

File: package.json (hunk 4 above)
Comment (tandrewnichols) [reply package.json]: Nm, I read the rollup doc on it.

**Comment by tandrewnichols** [comment]
It seems like [rollup-plugin-node-resolve](https://github.com/rollup/rollup-plugin-node-resolve) should've handled `import 'should'`, but I couldn't get it to work with any of the options it lists (`module: true` for example).

I know you've done a lot on this, but I want to pause and think about it. I'm not sure I want to incorporate rollup, since I'll be the one maintaining it, and frankly, my first experience was just not that good. Webpack may generate uglier and slightly bulkier code, but at least it knows how to handle `require('should')`. I'm also open to any other ideas that make the library more accessible for tools like rollup and webpack without such a wholesale change.

This isn't a "no." I just want to weigh the options.

**Comment by emilos** [comment]
hey @tandrewnichols no worries :) it took an hour or two only. Other option would be to just remove rollup/webpack altogether and simply write ES5 compatible library code by hand (which of course has downsides but does not require all this magic code :))

Any option is fine for me :) anything that seems simpler / more maintainable for you

**Comment by tandrewnichols** [comment]
What about the `browser` option in [node-resolve](https://github.com/rollup/rollup-plugin-node-resolve)? If I'm reading that correctly, that specifies an alternative file to load when bundling.

**Comment by emilos** [comment]
I didn't know about such option - sounds good to me!

File: karma.conf.js (hunk 1 above)
Comment (emilos) [reply karma.conf.js]: it's a recommended configuration which I found here: https://www.npmjs.com/package/karma-rollup-preprocessor

**Comment by tandrewnichols** [comment]
From reading, webpack supports the `browser` field, as does browserify and rollup (if you use the `rollup-plugin-node-resolve`). I'm gonna push up a branch with just that difference to ensure that it works as advertised. Do you want to try that branch out in your build to confirm rollup integration? I'm comfortable with webpack, so I'll try it there.

**Comment by tandrewnichols** [comment]
The branch is called `support-browser-builds`.

**Comment by emilos** [comment]
@tandrewnichols yup, that would work as well :) I'll close this one then :) thx!

**Pull request closed without merging**